### PR TITLE
Add missing cs_xval locks

### PR DESCRIPTION
**Pull request by kyuupichan**

### CODE REVIEW

File: src/unlimited.cpp [hunk 1]
```diff
@@ -1705,8 +1705,11 @@ extern UniValue getstructuresizes(const UniValue &params, bool fHelp)
 
     ret.push_back(Pair("mapBlockIndex", mapBlockIndex.size()));
     // CChain
-    ret.push_back(Pair("setPreVerifiedTxHash", setPreVerifiedTxHash.size()));
-    ret.push_back(Pair("setUnVerifiedOrphanTxHash", setUnVerifiedOrphanTxHash.size()));
+    {
+        LOCK(cs_xval);
+        ret.push_back(Pair("setPreVerifiedTxHash", setPreVerifiedTxHash.size()));
```
Comment (ptschip) [inline src/unlimited.cpp:1710]: i don't think you need this lock here, size() is thread safe.

Comment (kyuupichan) [reply src/unlimited.cpp:1710]: Is that guaranteed by the standard? I was thinking the implementation might have a pointer to an underlying object that returns the size and that that pointer might be reallocated by a modification in a separate thread.

Comment (kyuupichan) [reply src/unlimited.cpp:1710]: http://stackoverflow.com/questions/9022824/size-functions-and-thread-safety-in-c

says the lock is required

**Comment by ptschip** [comment]
@kyuupichan There are numerous places in the code (even right in that very same block) where size is accessed but not locked: It should be an an atomic operation.  I think the only time you have to worry about locking is when updating the same record concurrently, or adding/deleting while iterating ranges.


Here is what is written on size for c++ 98 regarding size
http://www.cplusplus.com/reference/map/map/size/
Size:
Data races
The container is accessed.
No elements are accessed: concurrently accessing or modifying them is safe.

**Comment by kyuupichan** [comment]
Thanks for the reference.  However reading closely I see it's talking about elements in the map and concurrent access to those. 

In particular, it isn't saying that adding or removing an element in a separate thread, which might cause the container to be reallocated, is safe.

I do agree this is likely safe for current implementations (and perhaps that it may always happen to be true), but I don't see the standard guaranteeing that even unlocked access to size() is safe (we don't care if it gives the wrong answer here - I agree, but we do care about size() accessing potentially invalid memory because the container resized), and I think we should be coding to the standard rather than implementations.

**Pull request merged**